### DISCUSSION TITLE
About Across Platform

### DISCUSSION THREAD

**Discussion by laizilianglaiziliang** [SpiceSharp/SpiceSharp#152] [General]
Hi, Sven. Spice# is a great work. Recently, I'm using Spice# for cross platform. I meet two problems.

First, I compared the two versions of c63a8ca of Feb 3 2019 with fd569ec of Feb 5 2019. The transistor of Feb 3 2019 can normally output transient simulation data after packaging cross platform files. But the Feb 5 2019 report “CircuitException: No parameter with the name 'is' found of type Double at SpiceSharp.ParameterSetDictionary .SetParameter[T]( System.String name, T value, System.Collections.Generic .IEqualityComparer`1[T] comparer)”

Second: I compared 3.0.X with 2.X. 3.0.X version can't output simulation data when cross platform, and the error "argumentnullexception: value cannot be null. Parameter name: format ”. However, in 2.X version, I did not encounter this problem.

I carefully looked at the questions raised in the project and compared the changes in various versions. I speculated that it was the impact of using reflection. I heard in an issuse that you want to cancel reflection,I'm looking forward to your unreflected version. I think it will better realize the cross platform function of spice#. I hope my question can be useful.

**Comment by svenboulanger** [comment]
Hi,

Thanks for the message :-) We have indeed seen some issues by going fully for reflection and inversion-of-control. The next version is in the making, and it uses source generators instead of reflection, which will allow better cross-platform support.

There are currently 4 (smaller) problems that I encountered doing so:
- Source generators support .NET Standard 2.0 and up, so the next version will drop support for most other target frameworks (I will just go for .NET Standard 2.0, as it is also recommended as the target framework for cross-platform applications).
- Development became a little bit more laggy. Source generators are very new, and it doesn't support incremental building yet (the source generator creates a lot of files every time it tries to build). This is not an issue for you guys, only for me and whoever wants to join.
- Because everything that previously used reflection is now hard-coded, the DLL size went up to ~1MB (at the time of writing). While this is in most cases not bad, it may have a bit of an impact when using the framework in online applications (I'm thinking Blazor).
- An extra project will need to be maintained now. The question is also how we can distribute this to users that are implementing their own custom models and components. If we decide to make it a NuGet package, then I feel we also need to guarantee some kind of robustness to errors and invalid code, as well as some diagnostic messages to help with development.

That being said, is this an issue that you see in particular frameworks? We know that reflection causes issues like this in AOT platforms, but if your platform isn't then maybe something else is going wrong?

**Reply by laizilianglaiziliang** [reply]
Thank you for your quick reply. Now I use .Net standard 2.0 and .Net 4.X for cross-platform development. After packaging my project, it runs perfectly in Windows and Android. However, the above problems will appear after packaging as a WebGL. I think it is the current webgl does not support reflection mechanism. 

Source code generator is too new, I know less. At present, below 2.7 version has been able to support my current project very well. But the running speed is slow, which I think it will be solved after you use the source code generator.

**Comment by laizilianglaiziliang** [comment]
I'm glad to see version 3.1.0. Now I can use resistor and capacitor to get the correct simulation data on different platforms, and the simulation speed has been significantly improved.

But when I use transistors, I still encounter the following error: "argumentnullexception: value cannot be null. Parameter name: format" System.String.FormatHelper ( System.IFormatProvider provider, System.String format, System.ParamsArray args)”.  I wonder if this is when the transistor is built, the reflection still exists.

**Reply by svenboulanger** [reply]
That looks like an error in `string.Format(...)`, so it doesn't look like reflection to me. Can you copy-paste the call stack that I can trace where it is calling the method `FormatString()`?
[edit] Or it could be an error thrown in a class that implements `IFormattable`.

**Comment by laizilianglaiziliang** [comment]
I'm not sure how to find the error now, but I uploaded my packaged test file to [https://github.com/laizilianglaiziliang/Spice-TestDemoBJT.git](url)

If you are interested, you can download and run tour_ testingserver.exe to open the program. The semi button is to test BJT, and the R button is to test resistance. You can press F12 to view the print data.

Here is my C# program before packing

`   

    void TestBJT()
    {
        /*
          * Transient analysis of a BJT common emitter amplifier
          * Output voltage should behave like the reference. Reference is from Spice 3f5.
          */
      
        // Create circuit
        var ckt = new Circuit(
            new VoltageSource("V1", "in", "0", new Pulse(0, 5, 1e-6, 1e-9, 0.5e-6, 2e-6, 6e-6)),
            new VoltageSource("Vsupply", "vdd", "0", 5.0),
            new Resistor("R1", "vdd", "out", 10.0e3),
            new Resistor("R2", "in", "b", 1.0e3),
            CreateBJT("Q1", "out", "b", "0", "0", "mjd44h11"),
                        CreateBJTModel("mjd44h11", string.Join(" ",
                "IS = 1.45468e-14 BF = 135.617 NF = 0.85 VAF = 10",
                "IKF = 5.15565 ISE = 2.02483e-13 NE = 3.99964 BR = 13.5617",
                "NR = 0.847424 VAR = 100 IKR = 8.44427 ISC = 1.86663e-13",
                "NC = 1.00046 RB = 1.35729 IRB = 0.1 RBM = 0.1",
                "RE = 0.0001 RC = 0.037687 XTB = 0.90331 XTI = 1",
                "EG = 1.20459 CJE = 3.02297e-09 VJE = 0.649408 MJE = 0.351062",
                "TF = 2.93022e-09 XTF = 1.5 VTF = 1.00001 ITF = 0.999997",
                "CJC = 3.0004e-10 VJC = 0.600008 MJC = 0.409966 XCJC = 0.8",
                "FC = 0.533878 CJS = 0 VJS = 0.75 MJS = 0.5",
                "TR = 2.73328e-08 PTF = 0 KF = 0 AF = 1"))
            );
        // Create simulation
        var tran = new Transient("tran", 1e-9, 10e-6);
        tran.ExportSimulationData += (sender, data) =>
        {
            Debug.Log(data.GetVoltage("out"));
        };
        tran.Run(ckt);
    }
`
`

    void TestR()
    {
        double dcVoltage = 10;
        var resistorResistance = 10e3; // 10000;
        var capacitance = 1e-6; // 0.000001;
        var tau = resistorResistance * capacitance;
        // Build circuit
        var ckt = new Circuit(
            new Capacitor("C1", "OUT", "0", capacitance),
            new Resistor("R1", "IN", "OUT", resistorResistance),
            new VoltageSource("V1", "IN", "0", dcVoltage)
            );
        // Create simulation, exports and references
        var tran = new Transient("tran", 1e-8, 10e-6);
        tran.ExportSimulationData += (sender, data) =>
        {
            Debug.Log(data.GetVoltage("OUT"));
        };
        tran.Run(ckt);
    }
`

**Reply by svenboulanger** [reply]
After googling your exception, it doesn't seem to be clear that it is Spice# that is causing that exception (a lot of SignalR + WebGL with different packages throw the exception).

Perhaps we should first make sure that the issue can be traced back to Spice#. Ideally, I'd like a unit test that demonstrates the problem.

**Reply by laizilianglaiziliang** [reply]
Thank you very much for your guidance and advice. During this time, I will check this problem and give feedback.
I guess there was a problem in configuring the transistor parameters. The previous 3.0. X version of the resistor will also encounter such errors. However, version 3.1 resistor devices can be used across-platforms. I will compare the configuration parameters of the transistor between c63a8ca and 3.1.

I believe SpiceSharp will become an outstanding cross-platform simulation library.

**Reply by svenboulanger** [reply]
If the problem is in the `string.Format()` method, then even informational and warning messages can cause issues. I.e. if the simulation should be able to run you will get an exception because it fails to display the warning.
